libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -364,14 +365,25 @@ void CodeTransform::operator()(Identifier const& _identifier)
 	if (m_scope->lookup(_identifier.name, Scope::NonconstVisitor(
 		[=](Scope::Variable& _var)
 		{
-			// TODO: opportunity for optimization: Do not DUP if this is the last reference
-			// to the top most element of the stack
-			if (int heightDiff = variableHeightDiff(_var, _identifier.name, false))
-				m_assembly.appendInstruction(dev::eth::dupInstruction(heightDiff));
+			decreaseReference(_identifier.name, _var);
+			if (int const heightDiff = variableHeightDiff(_var, _identifier.name, false))
+			{
+				if (m_variablesScheduledForDeletion.count(&_var))
+				{
+					if (heightDiff > 1)
+						m_assembly.appendInstruction(dev::eth::swapInstruction(heightDiff));
```

Comment:
I don't think this is correct, at least for `heightDiff >= 3` - you are swapping some random value into the stack.

I think the only two situations where this can be optimized are:

 - `heightDiff == 1` - replace `DUP1` by nothing (value is already there)
 - `heightDiff == 2` - replace `DUP1` by `SWAP1` (value is one slot further down)

Reply:
The following has a swap3:

```
object "Contract" {
    code {
        {
            let x := sload(0)
            let y := sload(1)
            mstore(y, 0)
            mstore(x, 0)
            mstore(y, 0)
        }
    }
}

Text representation:
    /* "test.yul":55:56   */
  0x00
    /* "test.yul":49:57   */
  sload
    /* "test.yul":82:83   */
  0x01
    /* "test.yul":76:84   */
  sload
    /* "test.yul":55:56   */
  0x00
    /* "test.yul":101:102   */
  dup2
    /* "test.yul":94:106   */
  mstore
    /* "test.yul":55:56   */
  0x00
    /* "test.yul":123:124   */
  swap3
    /* "test.yul":116:128   */
  mstore
    /* "test.yul":55:56   */
  0x00
    /* "test.yul":138:150   */
  mstore
```
Looks like the correct optimization for
```
    /* "test.yul":55:56   */
  0x00
    /* "test.yul":49:57   */
  sload
    /* "test.yul":82:83   */
  0x01
    /* "test.yul":76:84   */
  sload
    /* "test.yul":55:56   */
  0x00
    /* "test.yul":101:102   */
  dup2
    /* "test.yul":94:106   */
  mstore
    /* "test.yul":55:56   */
  0x00
    /* "test.yul":123:124   */
  dup3
    /* "test.yul":116:128   */
  mstore
    /* "test.yul":55:56   */
  0x00
    /* "test.yul":145:146   */
  dup2
    /* "test.yul":138:150   */
  mstore
  pop
  pop
```
(This passes all assertions)

Reply:
The version with a `swap3` is invalid. After the second `sload` the stack contains `x` and `y`, with `y` being on the top. Then you push `0`, so the stack contains `x y 0`. The code then requests `x`. One way to get it to the top would be a `swap2` (not a `swap3`). But both `swap2` and `swap3` also move the `0` down into the stack, although we need it right below the `x` for the `mstore` to work properly. In this case, either a much more involved mechanism or just the simple `dup3` would do the trick.

Reply:
Yes you're right, I get it now. Lets optimise `heightDIff <= 2` for now. I don't think we will have 1 opcode to do a `DUPx` and `POPx`.

Reply:
it should be `swapInstruction(1)`.